### PR TITLE
[Sparse] Migration of Unary, Softmax, Matmul Ops

### PR DESCRIPTION
**Pull request by mufeili**
## Description

I chose to re-use the matmul ops of `SparseMatrix` as much as possible for `DiagMatrix` as this is already a large PR and I want to meet the target DDL.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tests/pytorch/mock_sparse/test_mm.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+def test_sparse_dense_mm():
+    dev = F.ctx()
+    # A: shape (N, M), X: shape (M, F)
+    row = torch.tensor([0, 1, 1]).to(dev)
+    col = torch.tensor([1, 0, 1]).to(dev)
+    val = torch.randn(len(row)).to(dev)
+    A = create_from_coo(row, col, val)
+    X = torch.randn(2, 3).to(dev)
+    sparse_result = A @ X
+    dense_result = A.adj.to_dense() @ X
```
Comment (jermainewang) [inline tests/pytorch/mock_sparse/test_mm.py:16]: I see these tests are using the internal members of `A`. The issue is that when we replace the implementation in the future, we need to change test codes too. Can you directly use `row` and `col` to construct ground truth result?

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/mock_sparse/matmul.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+    val_dim = len(A2.val.shape)
+    assert val_dim == 1, f'Expect arg2.val to be a 1D tensor, got {val_dim}D'
+
+    if isinstance(A2, DiagMatrix):
+        return diag_diag(A1, A2)
+    else:
+        return sparse_sparse_mm(A1.as_sparse(), A2)
+
+def spmm(A1: Union[SparseMatrix, DiagMatrix], A2: Union[torch.Tensor, SparseMatrix, DiagMatrix])\
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/matmul.py]: Updated

File: python/dgl/mock_sparse/matmul.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        return torch.stack(A_list, dim=-1)
+    elif isinstance(A, SparseMatrix):
+        return create_from_coo(
+            row=A.row, col=A.col,
+            val=torch.stack([A_list[i].val for i in range(len(A_list))], dim=-1), shape=A.shape)
+    else:
+        return diag(
+            val=torch.stack([A_list[i].val for i in range(len(A_list))], dim=-1), shape=A.shape)
+
+def bspmm(A1: Union[SparseMatrix, DiagMatrix], A2: Union[torch.Tensor, SparseMatrix, DiagMatrix])\
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/matmul.py]: Updated

File: python/dgl/mock_sparse/matmul.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+    if A1.shape[1] != A2.shape[0]:
+        raise ValueError(f'Expect arg1.shape[1] and arg2.shape[0] to be the same, \
+                           got {A1.shape[1]} and {A2.shape[0]}')
+
+    if isinstance(A2, torch.Tensor):
+        return sparse_dense_mm(A1, A2)
+
+    val_dim = len(A2.val.shape)
+    assert val_dim == 1, f'Expect arg2.val to be a 1D tensor, got {val_dim}D'
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/matmul.py:175]: Updated a bit and see if it is better to you now. `if + raise` are used when we dispatch the input arguments to different underlying ops depending on the type of the input arguments.

File: python/dgl/mock_sparse/matmul.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+    if isinstance(A2, DiagMatrix):
+        A2 = A2.as_sparse()
+    return sparse_sparse_mm(A1, A2)
+
+def diag_diag(A1: DiagMatrix, A2: DiagMatrix) -> DiagMatrix:
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/matmul.py]: Changed to `diag_diag_mm`

**Review by jermainewang — APPROVED** [review]

**Comment by jermainewang** [comment]
The new logic is much cleaner. Nice job!

File: python/dgl/mock_sparse/matmul.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+
+    Examples
+    --------
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([1, 0, 1])
+    >>> val = torch.randn(len(row))
+    >>> A = create_from_coo(row, col, val)
+    >>> X = torch.randn(2, 3)
+    >>> result = A @ X
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse/matmul.py:149]: updated
